pysoa/server/action/switched.py

```diff
@@ -0,0 +1,126 @@
+        ):
+            raise ValueError(
+                'Class attribute switch_to_action_map must be an iterable of at least two indexable items, each '
+                'with exactly two indexes, where the first element is a switch and the second element is an action '
+                '(callable).'
+            )
+
+        return cls
+
+
+@six.add_metaclass(SwitchedActionMetaClass)
+class SwitchedAction(object):
+    """
+    A specialized action that defers to other, concrete actions based on request switches. Subclasses must not
+    override any methods and must override `switch_to_action_map`. `switch_to_action_map` should be some iterable
+    object that provides `__len__` (such as a tuple [recommended] or list). Its items must be indexable objects that
+    provide `__len__` (such as a tuple [recommended] or list) and have exactly two elements.
+
+    For each item in `switch_to_action_map`, the first element must be a switch that provides `__int__` (such as an
+    actual integer) or a switch that provides an attribute `value` which, itself, provides `__int__` (or is an int).
+    The second element must be an action, such as an action class (e.g. one that extends `Action`) or any callable
+    that accepts a server settings object and returns a new callable that, itself, accepts an `ActionRequest` object
+    and returns an `ActionResponse` object or raises an `ActionError`.
+
+    `switch_to_action_map` must have at least two items in it. `SwitchedAction` will iterate over that list, checking
+    the first element (switch) of each item to see if it is enabled in the request. If it is, the second element (the
+    action) of that item will be deferred to. If it finds no items whose switches are enabled, it will use the very
+    last action in `switch_to_action_map`. As such, you can treat the last item as a default, and its switch could
+    simply be `SwitchedAction.DEFAULT_ACTION` (although, this is not required: it could also be a valid switch, and
+    it would still be treated as the default in the case that no other items matched).
+
+    Example usage:
+
+    .. code-block:: python
+
+        class UserActionV1(Action):
+            ...
+
+        class UserActionV2(Action):
+            ...
+
+        class UserTransitionAction(SwitchedAction):
+            switch_to_action_map = (
+                (USER_VERSION_2_ENABLED, UserActionV2),
+                (SwitchedAction.DEFAULT_ACTION, UserActionV1),
+            )
+    """
+
+    DEFAULT_ACTION = -1827739192
+
+    switch_to_action_map = ()
+
+    def __init__(self, settings=None):
+        """
+        Construct a new action. Concrete classes should not override this.
+
+        :param settings: The server settings object
+        :type settings: dict
+        """
+        if self.__class__ is SwitchedAction:
+            raise TypeError('Cannot instantiate abstract SwitchedAction')
+
+        self.settings = settings
+
+    def get_raw_action(self, action_request):
```

Comment:
I'd call that `get_action_class` (even though in theory it can be something that is not a Python class, it's still used as such)

Reply:
I have renamed it `get_uninitialized_action`.